keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/Hamiltonian.java

```diff
@@ -28,185 +26,143 @@ public static NetworkSamples getPosteriorSamples(final BayesNet bayesNet,
-        final Map<String, Integer> indexLookup = createIndexLookup(bayesNet.getContinuousLatentVertices());
-        final MultivariateVectorFunction logOfMasterPGradientFunction = getLogOfMasterPGradientFunction(bayesNet);
+        final List<Vertex<Double>> latentVertices = bayesNet.getContinuousLatentVertices();
+        final List<Vertex<?>> probabilisticVertices = bayesNet.getVerticesThatContributeToMasterP();
+
+        final Map<String, List<?>> samples = new HashMap<>();
+        takeSamples(samples, fromVertices);
+
+        final Map<String, Double> positionBeforeLeapfrog = new HashMap<>();
+        final Map<String, Double> momentum = new HashMap<>();
+        final Map<String, Double> momentumBeforeLeapfrog = new HashMap<>();
 
         for (int sampleNum = 1; sampleNum < sampleCount; sampleNum++) {
 
-            final Map<String, Double> position = getLatestSample(samples);
-            final Map<String, Double> positionPreviousTimeStep = new HashMap<>(position);
+            cachePosition(latentVertices, positionBeforeLeapfrog);
 
-            final Map<String, Double> momentum = initializeMomentumForEachVertex(fromVertices, random);
-            final Map<String, Double> momentumPreviousTimeStep = new HashMap<>(momentum);
+            initializeMomentumForEachVertex(latentVertices, momentum, random);
+            cacheMomentum(momentum, momentumBeforeLeapfrog);
 
-            final double logOfMasterP = bayesNet.getLogOfMasterP();
+            final double logOfMasterPBeforeLeapfrog = bayesNet.getLogOfMasterP();
+
+            Map<String, Double> gradient = LogProbGradient
+                    .getJointLogProbGradientWrtLatents(bayesNet.getVerticesThatContributeToMasterP());
 
             for (int leapFrogNum = 0; leapFrogNum < leapFrogCount; leapFrogNum++) {
-                leapfrogVertices(
-                        fromVertices,
-                        position,
+                gradient = leapfrog(
+                        latentVertices,
+                        gradient,
                         momentum,
                         stepSize,
-                        bayesNet,
-                        logOfMasterPGradientFunction,
-                        indexLookup
+                        probabilisticVertices
                 );
             }
 
             final double likelihoodOfLeapfrog = getLikelihoodOfLeapfrog(
                     bayesNet,
                     momentum,
-                    logOfMasterP,
-                    momentumPreviousTimeStep
+                    logOfMasterPBeforeLeapfrog,
+                    momentumBeforeLeapfrog
             );
 
-            if (shouldAccept(likelihoodOfLeapfrog, random)) {
-                addAsLatest(samples, position);
-            } else {
-                addAsLatest(samples, positionPreviousTimeStep);
+            if (shouldReject(likelihoodOfLeapfrog, random)) {
+                revertToPosition(latentVertices, positionBeforeLeapfrog);
```

Comment:
As mentioned this morning, what do you think would work best?

1. Reverting the values by recalculating them
2. Store a 'cached' version of the BayesNet itself and swap it back on a reject